semantic/type_inference.md

```diff
@@ -0,0 +1,138 @@
+### Type Inference in Flux
+--------------------------
+
+Flux is a statically typed language.
+Flux does not require explicit type annotions but rather infers the most general type of a variable.
+Flux supports parametric polymorphism.
+
+#### Rules
+----------
+
+In order to perform type inference Flux has a collection of rules that govern the relationship between types and expressions.
+Take the following program for example.
+
+```
+f(a, b) => return a + b
+x = f(a:1, b:2)
+```
+
+The following rules will be used to solve for the types in the above program:
+
+1. Binary Addition Rule
+
+    typeof(a + b) = typeof(a)  
+    typeof(a + b) = typeof(b)  
+
+2. Function Rule
+
+    typeof(f) = [ typeof(a), typeof(b) ] -> typeof(return statement)
+
+3. Call Expression Rule
+
+    typeof(f) = [ typeof(a), typeof(b) ] -> typeof( f(a:1, b:2) )
+
+4. Variable Declaration Rule
+
+    typeof(x) = typeof( f(a:1, b:2) )
+
+#### Algorithm
+--------------
+
+The type inference algorithm that Flux uses is based on Wand's algorithm.
+It is completed in the following 3 phases:
+
+1. Type annotation phase
+
+    Input: Semantic graph  
+    Output: Map of expressions to type variables  
+
+    In this phase every expression is assigned a type variable which is a placeholder for a concrete type.
+
+    Example:
+
+        Program:
+        
+            x = 1 + 2
+
+        Output Map:
+
+            x   -> t0
+            1+2 -> t1
+            1   -> t2
+            2   -> t3
+
+2. Constraint generation phase
+
+    Input: Map of expressions to type variables  
+    Output: Map of type variables to type expressions  
```

Comment:
I expected the output to be a set of constraints not a map. Below your example shows it as a set.